### PR TITLE
Revamp docs

### PR DESCRIPTION
**Pull request by thomaspinder**
This PR provides a major update to the docs. Changes include moving the backend to MKDocs, updating docstrings, and splitting out notebooks.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [X] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

The current docs are outdated.

### CODE REVIEW

File: gpjax/linops/constant_diagonal_linear_operator.py [hunk 1]
```diff
@@ -83,62 +90,66 @@ def __mul__(self, other: float) -> LinearOperator:
...
+        Returns
+        -------
             LinearOperator: Covariance operator multiplied by a scalar.
         """
-
         return ConstantDiagonalLinearOperator(value=self.value * other, size=self.size)
 
     def _add_diagonal(self, other: DiagonalLinearOperator) -> LinearOperator:
-        """Add diagonal to the covariance operator,  useful for computing, Kxx + Iσ².
+        """Add diagonal to the covariance operator,  useful for computing, Kxx + Io².
```
Comment (st--) [inline gpjax/linops/constant_diagonal_linear_operator.py:100]:
```suggestion
        """Add diagonal to the covariance operator,  useful for computing, Kxx + Iσ².
```

File: gen_pages.py
Comment (st--) [inline gen_pages.py:1]: is this a duplicate of docs/scripts/gen_pages.py (at least it visually looks similar), should it be deleted?

File: gpjax/__init__.py [hunk 2]
```diff
@@ -19,7 +19,28 @@
...
+    Matern32,
+    Matern52,
+    Periodic,
+    Polynomial,
+    PoweredExponential,
+    ProductKernel,
+    RationalQuadratic,
+    SumKernel,
+    White,
+)
```
Comment (st--) [inline gpjax/__init__.py:43]: i'm assuming this is some kind of auto-expander? this is one of those rare cases where I'd prefer the `import *` syntax...

File: gpjax/base/module.py [hunk 3]
```diff
@@ -56,7 +54,7 @@
 
 
 class Module(Pytree):
-    _pytree__meta: Dict[str, Any] = static_field()
+    _pytree__meta: dict[str, Any] = static_field()
```
Comment (st--) [inline gpjax/base/module.py:57]:
```suggestion
    _pytree__meta: Dict[str, Any] = static_field()
```

File: gpjax/base/param.py [hunk 4]
```diff
@@ -21,22 +21,21 @@
 from beartype.typing import (
     Any,
     Mapping,
```
Comment (st--) [inline gpjax/base/param.py:23]:
```suggestion
    Mapping,
    Optional,
```

File: gpjax/base/param.py [hunk 5]
```diff
@@ -21,22 +21,21 @@
...
-    Optional,
 )
 import tensorflow_probability.substrates.jax.bijectors as tfb
 
 
 def param_field(
     default: Any = dataclasses.MISSING,
     *,
-    bijector: tfb.Bijector = tfb.Identity(),
+    bijector: tfb.Bijector = None,
```
Comment (st--) [inline gpjax/base/param.py:31]:
```suggestion
    bijector: Optional[tfb.Bijector] = None,
```

File: gpjax/base/param.py [hunk 6]
```diff
@@ -21,22 +21,21 @@
...
     trainable: bool = True,
     default_factory: Any = dataclasses.MISSING,
     init: bool = True,
     repr: bool = True,
-    hash: Optional[bool] = None,
+    hash: bool | None = None,
     compare: bool = True,
-    metadata: Optional[Mapping[str, Any]] = None,
+    metadata: Mapping[str, Any] | None = None,
```
Comment (st--) [inline gpjax/base/param.py:38]:
```suggestion
    metadata: Optional[Mapping[str, Any]] = None,
```

File: gpjax/fit.py [hunk 7]
```diff
@@ -115,6 +116,8 @@ def fit(
         Tuple[Module, Array]: A Tuple comprising the optimised model and training
             history respectively.
     """
+    if key is None:
+        key = jr.PRNGKey(42)
```
Comment (st--) [inline gpjax/fit.py:120]: seems kinda dangerous pre-specifying a fixed seed...

File: gpjax/gaussian_distribution.py [hunk 8]
```diff
@@ -76,7 +75,7 @@ class GaussianDistribution(tfd.Distribution):
     """Multivariate Gaussian distribution with a linear operator scale matrix.
 
     Args:
-        loc (Optional[Float[Array, "N"]]): The mean of the distribution. Defaults to None.
+        loc (Optional[Float[Array, " N"]]): The mean of the distribution. Defaults to None.
```
Comment (st--) [inline gpjax/gaussian_distribution.py]:
```suggestion
        loc (Optional[Float[Array, "N"]]): The mean of the distribution. Defaults to None.
```

File: tests/test_kernels/test_utils.py [hunk 9]
```diff
@@ -24,6 +27,12 @@
 
 from gpjax.kernels.stationary.utils import euclidean_distance
 
+if TYPE_CHECKING:
+    from jaxtyping import (
+        Array,
+        Float,
+    )
+
```
Comment (st--) [inline tests/test_kernels/test_utils.py:35]:
```suggestion
```

File: tests/test_kernels/test_utils.py [hunk 10]
```diff
@@ -13,7 +13,10 @@
 # limitations under the License.
 # ==============================================================================
 
-from typing import List
+from typing import (
+    TYPE_CHECKING,
+    List,
+)
```
Comment (st--) [inline tests/test_kernels/test_utils.py:19]:
```suggestion
from typing import List
```

File: tests/test_kernels/test_stationary.py [hunk 11]
```diff
@@ -123,9 +123,6 @@ def test_gram(self, dim: int, n: int) -> None:
         # Initialise kernel
         kernel: AbstractKernel = self.kernel()
 
-        # Gram constructor static method
-        kernel.gram
```
Comment (st--) [inline tests/test_kernels/test_stationary.py:127]: was this supposed to get called?

File: tests/test_kernels/test_approximations.py [hunk 12]
```diff
@@ -31,7 +31,7 @@
 @pytest.mark.parametrize("num_basis_fns", [2, 10, 20])
 @pytest.mark.parametrize("n_dims", [1, 2, 5])
 def test_frequency_sampler(kernel: AbstractKernel, num_basis_fns: int, n_dims: int):
-    key = jr.PRNGKey(123)
+    jr.PRNGKey(123)
```
Comment (st--) [inline tests/test_kernels/test_approximations.py:34]:
```suggestion
```
doesn't serve any purpose?

File: tests/test_gaussian_distribution.py [hunk 13]
```diff
@@ -47,7 +46,7 @@ def test_array_arguments(n: int) -> None:
     mean = jr.uniform(key_mean, shape=(n,))
     sqrt = jr.uniform(key_sqrt, shape=(n, n))
     covariance = sqrt @ sqrt.T
-    L = jnp.linalg.cholesky(covariance)
+    jnp.linalg.cholesky(covariance)
```
Comment (st--) [inline tests/test_gaussian_distribution.py:49]:
```suggestion
    _L = jnp.linalg.cholesky(covariance)  # check that cholesky does not error
```
?

File: gpjax/variational_families.py [hunk 14]
```diff
@@ -638,7 +640,6 @@ def predict(
         # AAᵀ
         AAT = jnp.matmul(A, A.T)
 
-        # LLᵀ = I + AAᵀ
```
Comment (st--) [inline gpjax/variational_families.py:641]: I liked the comment...

File: README.md [hunk 15]
```diff
@@ -125,7 +125,8 @@ opt_posterior, history = gpx.fit(
     train_data=D,
     optim=optimiser,
     num_iters=500,
-    safe=True
+    safe=True,
+    key = key
```
Comment (st--) [inline README.md:129]:
```suggestion
    key=key,
```
to match with black formatting?

File: docs/examples/regression.ipynb
Comment (st--) [inline docs/examples/regression.ipynb:1]: should this be removed (in favour of the jupytext version below)?

File: docs/index copy.md
Comment (st--) [inline docs/index copy.md:1]: accidental copy - remove?

File: gpjax/base/module.py [hunk 16]
```diff
@@ -202,6 +200,16 @@ def _apply_stop_grad(meta_leaf):
 
         return meta_map(_apply_stop_grad, self)
 
+    def trainables(self) -> Self:
```
Comment (st--) [inline gpjax/base/module.py:203]:
```suggestion
    def trainables(self: Self) -> Self:
```
otherwise "Self" doesn't contain any information

File: docs/examples/README.md [hunk 17]
```diff
@@ -60,17 +67,18 @@ class Prior(AbstractPrior):
     Example:
         >>> import gpjax as gpx
         >>>
+        >>> meanf = gpx.mean_functions.Zero()
         >>> kernel = gpx.kernels.RBF()
-        >>> prior = gpx.Prior(kernel = kernel)
+        >>> prior = gpx.Prior(mean_functions=meanf, kernel = kernel)
```
Comment (st--) [inline docs/examples/README.md:72]:
```suggestion
        >>> prior = gpx.Prior(mean_function=meanf, kernel = kernel)
```
?

File: docs/index copy.md
Comment (st--) [inline docs/index copy.md:1]: delete ?

File: gpjax/gps.py [hunk 18]
```diff
@@ -460,46 +489,52 @@ def sample_approx(
...
+                [Wilson et. al. (2020)](https://arxiv.org/abs/2002.09309) for expressions
+                for the canonical weights $`v_j`$.
+        =======
+                $\hat{f}(x) = \sum_{i=1}^m \phi_i(x)\theta_i + \sum{j=1}^N v_jk(.,x_j)$
+                where $\phi_i$ are m features sampled from the Fourier feature decomposition of
+                the model's kernel and $k(., x_j)$ are N canonical features. The Fourier
+                weights $\theta_i$ are samples from a unit Gaussian. See
+                [Wilson et. al. (2020)](https://arxiv.org/abs/2002.09309) for expressions
+                for the canonical weights $v_j$.
+        >>>>>>> 40b9a3b23c487221dacde900ad79e4d94a1c5d17
```
Comment (st--) [inline gpjax/gps.py:520]: looks like a missed merge-conflict.

btw, why do the inline maths now have to be ``$` ... `$``?

Comment (thomaspinder) [reply gpjax/gps.py:520]: The dollar dash syntax is how katex renders maths.
